src/tentaclio/clients/gcs_client.py

```diff
@@ -0,0 +1,160 @@
+"""GCS Stream client."""
+from typing import Optional, Union, Tuple, cast
+
+from google.auth import credentials, default
+from google.cloud import storage, exceptions as google_exceptions
+
+from tentaclio import urls, protocols
+
+from . import base_client, decorators, exceptions
+
+
+__all__ = ["GCSClient"]
+
+
+class GCSClient(base_client.BaseClient["GCSClient"]):
+    """GCS client.
+
+    Ref: https://googleapis.dev/python/storage/latest/index.html
+    """
+
+    allowed_schemes = ["gcs"]
+
+    conn: storage.Client
+    project: Optional[str] = None
+    creds: Optional[credentials.Credentials] = None
+    bucket: Optional[str]
+    key_name: Optional[str]
+
+    def __init__(
+        self,
+        url: Union[urls.URL, str],
+        project: Optional[str] = None,
+        creds: Optional[credentials.Credentials] = None
```

Comment:
Passing the credentials like this isn't really canonical tentaclio, because we'd prefer to have the primary method of authentication to be embedded in the URL. This is what we do in S3 as well - but in practice, no one stores their S3 credentials in tentaclio but rather they've already logged in to the aws cli, which stores the credentials in a default place that boto (the s3 client library) knows how to read.

So, this will in practice do the same thing: the google storage client will read the credentials from the default place (and in production, the file will be placed or an environment variable set or whatever). However it'd be nice to be able to authenticate the "usual" way as well. I had a look and google uses OAuth2, unfortunately requiring an id, a secret _and_ a refresh token, and I'm not sure how that would fit into the URI authority format.

Default behaviour I think is to consider anything after the `:` and before the `@` as secret in `scheme:user:pass@host:port/path`, so we could shoehorn both tokens in there with a separator or something... but maybe thats too much of a hack. Javi might have some thoughts or he might just think its chillin to leave the auth to the google client library. Maybe when you make the PR to the main repo, add a line in the description about the google auth situation

Reply:
[See comment](https://github.com/datavaluepeople/tentaclio/pull/1#issuecomment-635339555)